Remove notifications by their key alone

JuneKelly opened the pull request.

Commit:
add a .nvmrc file

Commit:
Add an endpoint to remove a notification by its key alone

Commit:
Add an optional expiry to notifications.

henryoswald commented on app.coffee:
```diff
@@ -24,6 +24,7 @@ app.post '/user/:user_id', controller.addNotification
 app.get '/user/:user_id', controller.getUserNotifications
 app.del '/user/:user_id/notification/:notification_id', controller.removeNotificationId
 app.del '/user/:user_id', controller.removeNotificationKey
+app.del '/notification/key/:key', controller.removeNotificationByKeyOnly
```
do we need /notification in the url here?

JuneKelly replied:
perhaps not, the odds of `notifications` ever managing another object aside from Notification are slim :)

henryoswald commented on app/coffee/Notifications.coffee:
```diff
@@ -22,25 +22,36 @@ module.exports =
+					expires: notification.expires || false
+				# ttl index on `expiresFrom` field
+				if doc.expires
+					doc.expiresFrom = new Date()
```
assume we can't put the expire index on the doc.expires? if so little comment explaining why we have doc.expires and doc.expiresFrom would be good

JuneKelly replied:
doc.expires would be a boolean, indicating that the doc should expire, while expiresFrom would be the actual ttl'd date column. TTL indexes in Mongo only work on Date fields, will add a comment to that effect

Commit:
Add commentary on ttl index

henryoswald commented on app/coffee/Notifications.coffee:
```diff
@@ -22,25 +22,38 @@ module.exports =
+				# in Mongo, TTL indexes only work on date fields,
+				# and ignore the document when that field is missing
+				if doc.expires
+					doc.expiresFrom = new Date()
```
what if we did

```
if notification.expires?
    doc.expires = notification.expires
```

rather than having expires set to false

Commit:
Merge branch 'master' of github.com:sharelatex/notifications-sharelatex

Commit:
Merge branch 'master' into pr-remove-by-key

# Conflicts:
#	test/unit/coffee/NotificationsTests.coffee

Commit:
Alter how the notification doc is built

Commit:
Remove redundant `/notification` path component

Commit:
Change how notification expiry works

Set the `expires` field to be a date in the future, after which the
document should be removed.

henryoswald commented on app/coffee/Notifications.coffee:
```diff
@@ -22,25 +22,38 @@ module.exports =
+				# in Mongo, TTL indexes only work on date fields, and ignore the document when that field is missing
+				# see `README.md` for instruction on creating TTL index
+				if notification.expires?
+					doc.expires =  new Date(notification.expires)
```
pedantic but worth a try catch?

JuneKelly replied:
Good catch, will add.

Commit:
return an error if `expires` is not a valid date.
